### PR TITLE
[ENH] Clumper write csv

### PR DESCRIPTION
**Pull request by samukweku**
This function writes a `Clumper` to csv. It also takes care of null values/missing keys.

An edit was made to read_csv to accommodate missing keys and data types.

### CODE REVIEW

File: clumper/clump.py [hunk 1]
```diff
@@ -174,19 +177,112 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
...
                 raise ValueError("Number of lines to read must be > 0.")
 
-        if path.startswith("https:") or path.startswith("http:"):
+        # conveniently excludes pathlib files here
+        # and removes the need to write code to check pathlib files in other places
+        if isinstance(path, str) and path.startswith(("https:", "http:")):
             with urllib.request.urlopen(path) as resp:
                 if fieldnames is None:
                     fieldnames = resp.readline().decode().strip().split(",")
                 # this section allows us to chunk the rows, if nrows is supplied
```
Comment (koaning) [inline clumper/clump.py:186]: this -> This

File: clumper/clump.py [hunk 2]
```diff
@@ -174,19 +185,112 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
...
-                return Clumper([dict(zip(key, values)) for key, values in body])
+                result = [dict(zip(key, values)) for key, values in body]
+        else:
+            with open(path, newline="") as csvfile:
+                reader = csv.DictReader(
+                    csvfile, delimiter=delimiter, fieldnames=fieldnames
+                )
+                reader = csv.DictReader(
+                    csvfile, delimiter=delimiter, fieldnames=fieldnames
+                )
```
Comment (koaning) [inline clumper/clump.py:209]: This feels like like repetition.

File: clumper/clump.py [hunk 3]
```diff
@@ -174,19 +185,112 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
             if n <= 0:
                 raise ValueError("Number of lines to read must be > 0.")
 
-        if path.startswith("https:") or path.startswith("http:"):
+        # conveniently excludes pathlib files here
+        # and removes the need to write code to check pathlib files in other places
+        if isinstance(path, str) and path.startswith(("https:", "http:")):
```
Comment (koaning) [inline clumper/clump.py:190]: The comments in general deserve a bit more punctuation.

File: clumper/clump.py [hunk 4]
```diff
@@ -174,19 +185,112 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
...
+                # python version less than 3.8 returns an OrderedDict
+                result = [dict(entry) for entry in it.islice(reader, 0, n)]
+
+        # null values, same as missing keys.
+        # if there are null values/missing keys, they will be truncated from the dictionary
+        # python's csv module treats null values as empty strings when writing to a csv -
+        # https://docs.python.org/3.8/library/csv.html#csv.DictWriter
+        # the user can choose to explicitly show missing keys/null values in the dictionary
+        # by assigning `ignore` to the na_values argument. At the moment, the default for
+        # null values are empty string ("") and "NA".
```
Comment (koaning) [inline clumper/clump.py:219]: The comments in general deserve a bit more punctuation.

File: tests/test_read_write/test_write_csv.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    Clumper.read_csv("tests/data/monopoly.csv").write_csv(path)
+    reader = Clumper.read_csv(path)
+    assert Clumper.read_csv("tests/data/monopoly.csv").collect() == reader.collect()
+
+
+def test_write_csv_append_mode(tmp_path):
+    """Test that the append mode works."""
+    d = tmp_path / "sub"
+    d.mkdir()
+    path = d / "monopoly_copy.csv"
```
Comment (koaning) [inline tests/test_read_write/test_write_csv.py:18]: This bit of code repeats itself in these tests. Might be good to write a helper function to create this pre-existing path.

File: tests/test_read_write/test_write_csv.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    # the header is appended as well, hence the +1
+    assert len(reader) == (2 * len(Clumper.read_csv("tests/data/monopoly.csv"))) + 1
+
+
+@pytest.fixture()
+def data():
+    return [
+        {"One": "a", "Two": "1", "Three": "one"},
+        {"One": "b", "Two": "2", "Three": "two"},
+        {"One": "", "Two": "3", "Three": "three"},
```
Comment (koaning) [inline tests/test_read_write/test_write_csv.py:31]: Could we add an example like: 

```
{"Two": 3, "Three": "three"}
```

That feels like an interesting edge case to make sure that we get right.

File: tests/test_read_write/test_write_csv.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    d.mkdir()
+    path = d / "nulls.csv"
+    Clumper(data).head(n=10).write_csv(path)
+    reader = Clumper.read_csv(path, na_values="ignore")
+    assert Clumper(data).head(n=10).collect() == reader.collect()
+
+
+def test_write_missing_keys(tmp_path):
+    """Test that function works with missing keys."""
+    data2 = [{"a": "1", "b": "2"}, {"c": "3"}]
```
Comment (koaning) [inline tests/test_read_write/test_write_csv.py:71]: There's always a better variable name than `data2`.

File: clumper/clump.py [hunk 8]
```diff
@@ -147,6 +150,14 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
                         the correct number of fieldnames is supplied, as an incorrect number can lead
                         to truncation. If you have seven columns and your fieldnames length is 3,
                         then every row will have only 3 values, the remaining four will be cut off.
+            na_values:  Null values. If `ignore`, null values are passed as is. If `None`, then for each row,
```
Comment (koaning) [inline clumper/clump.py:153]: > null values are passed as is 

That means we'll get empty strings or `None` values? Might be good to emphasize that.

**Comment by koaning** [comment]
I certainly appreciate the effort, lot's of work has gone in, thanks for that! 

That said, there's a larger topic to discuss in this PR. The doc-string is a little bit confusing to me at the moment. I noticed that I don't fully understand what the `fieldnames` and `na_values` arguments do. It might help to add a test or two that focus on just making it clear how these arguments change the output on disk.

File: clumper/clump.py [hunk 9]
```diff
@@ -174,19 +185,112 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
...
+        from clumper import Clumper
+
+        from tempfile import TemporaryDirectory
+        from pathlib import Path
+        with TemporaryDirectory() as temporary_dir:
+            Path(temporary_dir).touch('monopoly_copy.csv')
+            path = Path(temporary_dir)/'monopoly.csv'
+            Clumper.read_csv("tests/data/monopoly.csv").write_csv(str(path))
+            reader = Clumper.read_csv(str(path))
+            assert Clumper.read_csv("tests/data/monopoly.csv").collect() == reader.collect()
```
Comment (koaning) [inline clumper/clump.py:282]: Since this is a docstring example it should suffice to just add `/tmp/monopoly.csv` as a filepath.

File: clumper/clump.py (hunk 2 above)
Comment (samukweku) [reply clumper/clump.py:209]: Thanks for this. Missed it. I'll have a look at it today

File: tests/test_read_write/test_write_csv.py (hunk 6 above)
Comment (samukweku) [reply tests/test_read_write/test_write_csv.py:31]: The only way with this is to that all values assigned to two are either strings or numbers, not a mix. If it is a mix, then it comes back as a string type.

**Comment by koaning** [comment]
@samukweku is it ready for another review? Feel free to ping me if so.

**Comment by samukweku** [comment]
Oh. Ok. I made some commits. Have a look

File: clumper/clump.py [hunk 10]
```diff
@@ -174,19 +188,102 @@ def read_csv(cls, path, delimiter=",", fieldnames=None, n=None):
...
+        else:
+            with open(path, newline="") as csvfile:
+                reader = csv.DictReader(
+                    csvfile, delimiter=delimiter, fieldnames=fieldnames
+                )
+                # python version less than 3.8 returns an OrderedDict
+                result = [dict(entry) for entry in it.islice(reader, 0, n)]
+
+        # null values, same as missing keys.
+        # if there are null values/missing keys, they will be truncated from the dictionary.
```
Comment (koaning) [inline clumper/clump.py:211]: Nitpick. Punctuation is missing -> needs to start with captical letters.

**Review by koaning — APPROVED** [review]

**Comment by koaning** [comment]
I've played around with it and it seems to work like I expect it. Thanks! 

I've felt a small comment on punctuation in the comments. It is a nitpick, but a final thing you could address.

File: clumper/clump.py (hunk 10 above)
Comment (samukweku) [reply clumper/clump.py:211]: Hahaha. Sure. I'll fix it.

Comment (samukweku) [reply clumper/clump.py]: @koaning made the punctuation updates
